File: apps/mobile/modules/import-sources/ios/ProgressThrottle.swift
import Foundation

/// Progress-event gate: emit when at least 100ms elapsed or at least 1MB
/// moved since the last emission. The first update always passes. The last
/// throttled event may under-report; completion is signaled by the copy's
/// result, not by a final progress event.
public struct ProgressThrottle {
  public static let minInterval: TimeInterval = 0.1
  public static let minBytes: Int64 = 1_048_576

  private var lastBytes: Int64 = -1
  private var lastTime: TimeInterval = -1

  public init() {}

  public mutating func shouldEmit(bytes: Int64, now: TimeInterval) -> Bool {
    if lastTime < 0 || now - lastTime >= Self.minInterval || bytes - lastBytes >= Self.minBytes {
      lastBytes = bytes
      lastTime = now
      return true
    }
    return false
  }
}

File: apps/mobile/modules/import-sources/android/src/main/java/expo/modules/importsources/ProgressThrottle.kt
package expo.modules.importsources

/**
 * Gates progress events: emit when at least 100ms elapsed or 1MB moved since
 * the last emission. The first update always passes; the final emission is
 * the caller's job because the throttle cannot know which update is last.
 */
class ProgressThrottle {
  companion object {
    const val MIN_INTERVAL_MS = 100L
    const val MIN_BYTES = 1_048_576L
  }

  private var lastBytes = -1L
  private var lastTimeMs = -1L

  fun shouldEmit(bytes: Long, nowMs: Long): Boolean {
    if (lastTimeMs < 0 || nowMs - lastTimeMs >= MIN_INTERVAL_MS || bytes - lastBytes >= MIN_BYTES) {
      lastBytes = bytes
      lastTimeMs = nowMs
      return true
    }
    return false
  }
}
